### PR TITLE
Make it easier to use area detectors with a factory

### LINKED ISSUE #984: area detector factory function
A factory function might make it easier to create area detector classes and instances.  The factory would address common configurations (such as PVA, HDF5, ...) as options, triggered by keywords.  Additional kwargs would describe supplemental options such as the file path seen by the IOC.

### PR DESCRIPTION
**Pull request by prjemian**
- close #984

### CODE REVIEW

**Comment by prjemian** [comment]
The initial commit is for a fixed version of ophyd area detector support.  It is templated on the support provided for the APS XPCS instrument.

**Comment by prjemian** [comment]
Still need to add/blend [contribution](https://github.com/BCDA-APS/apstools/issues/984#issuecomment-2192536006) by @cooleyv that requests the ADCore version from the existing PV and [use case](https://github.com/BCDA-APS/apstools/issues/984#issuecomment-2195201893) from @canismarko

The challenge with asking the ADCore version is that the IOC must be running enough to respond.

**Comment by prjemian** [comment]
Will be a fun challenge blending these and then building unit tests.

File: apstools/devices/area_detector_support.py [hunk 1]
```diff
@@ -825,6 +826,186 @@ def unstage(self):
...
+    * ``generate_datum()`` - coordinate image storage metadata
+    """
+
+    def stage(self):
+        self.stage_sigs.move_to_end("capture", last=True)
+        super().stage()
+
+
+def ad_class_factory(
+    cam_class,
```
Comment (prjemian) [inline apstools/devices/area_detector_support.py:860]: Change to kwarg, default to CamBase

**Comment by prjemian** [comment]
The factory method looks too complex.  Can it be simpler?  ([advice](https://www.geeksforgeeks.org/class-factories-a-powerful-pattern-in-python/))

**Comment by prjemian** [comment]
There are very few detectors in use at APS that still use ADCore releases older than 3.4.  Additional support for versions of the various plugins is not justified.  They can be addressed individually if the new factory design is not adequate.

**Comment by prjemian** [comment]
Needs some tests for [`instantiate_fake_device()`](https://github.com/BCDA-APS/apstools/issues/984#issuecomment-2195201893), as suggested by @canismarko.

**Comment by prjemian** [comment]
Instead of: https://github.com/BCDA-APS/apstools/blob/abd00fe59fa63170b2dc6e6b3161a4499bbe1827/apstools/devices/area_detector_factory.py#L147-L148
to remove a default, such as: https://github.com/BCDA-APS/apstools/blob/abd00fe59fa63170b2dc6e6b3161a4499bbe1827/apstools/devices/area_detector_factory.py#L53-L60

allow the caller to re-define a key with `defaults` kwarg.  This kwargs will update the `PLUGIN_DEFAULTS` dictionary: https://github.com/BCDA-APS/apstools/blob/abd00fe59fa63170b2dc6e6b3161a4499bbe1827/apstools/devices/area_detector_factory.py#L82

**Comment by prjemian** [comment]
@cooleyv, @keenanlang, @canismarko, @gfabbris, @strempfer, @MarkRivers  -- The new area detector factory support is ready for review.

**Comment by prjemian** [comment]
@sureshnaps : Your suggestion has led to simplifications such as.

```py
    from ophyd.areadetector import SimDetectorCam
    from apstools.devices import ad_creator

    det = ad_creator(
        "ad:", name="det", class_name="MySimDetector",
        [
            {"cam": {"class": SimDetectorCam}},
            "image",  # all the defaults are acceptable
            "pva",
            {"hdf1": {"write_path_template": "/", "read_path_template": "/mnt/ioc/tmp"}},
            "codec1",
            "proc1",
            "roi1",
            "stat1",
        ],
    )
```

**Comment by prjemian** [comment]
In the `PLUGIN_DEFAULTS`, the origin (ophyd _v_. apstools) of the default class for each plugin should be more obvious.  Use absolute paths for the classes from ophyd.

**Review by MDecarabas — APPROVED** [review]
Looks good, can you make an issue to remove the local version out of 8-ID and transition to the APS tools version. We can make the changes at the same time we re-implement specwriter

File: apstools/devices/tests/test_issue984_ad_factory.py
Comment (canismarko) [inline apstools/devices/tests/test_issue984_ad_factory.py:1]: This pretty much captures how I use this feature. I'll keep an eye out for edge cases where this doesn't work, but I can't think of any right now.

**Review by canismarko — APPROVED** [review]
This strikes me as a positive change.

The difference between the *plugins* and *plugin_defaults* arguments to the AD factory is tricky for me to understand. I think this is because of the complicated way area detectors are built in ophyd rather than a design issue for the factory, so probably this needs to be well documented.

File: apstools/devices/tests/test_issue984_ad_factory.py
Comment (prjemian) [reply apstools/devices/tests/test_issue984_ad_factory.py:1]: I auditioned a few names for `plugin_defaults` before deciding on a name which matched the `PLUGIN_DEFAULTS` in the module.  The general idea here, naming aside, is that the default plugin classes decided by the module from apstools could be replaced completely by the caller.  One advantage is future-proofing, as newer versions of the support classes appear.  Another case that came up in testing directly was how to remove the `write_path_template` key from `hdf1`.  It seems more natural to be able to edit or replace the `plugin_defaults` dictionary.

Documentation should clarify this.

Might better names for `plugins` and `plugin_defaults` help to distinguish their purpose?  I'm thinking `config` instead of `plugins`, for example.

Comment (canismarko) [reply apstools/devices/tests/test_issue984_ad_factory.py:1]: I don't have better suggestions for the naming. I think the name *plugins* does accurately reflect what the argument is doing, since in its simplest use it really is just a list of plugins to include. I think keeping *plugin_defaults* to match ``PLUGIN_DEFAULTS`` makes sense too.

One easy change might be to use keyword arguments in the example at the top of `area_detector_factory.py` so it's clear what is *plugins* and what is *plugin_defaults*, at least for the example that includes both. On first read I got lost in the curly and square brackets, making it harder to understand.

File: apstools/devices/area_detector_factory.py [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+   ~ad_creator
+   ~ad_class_factory
+   ~PLUGIN_DEFAULTS
+
+EXAMPLES
+
+Just the camera plugin (uses CamBase, the most basic features)::
+
+    from apstools.devices import ad_creator
+    det = ad_creator("ad:", name="det", class_name="MySimpleAD", ["cam",])
```
Comment (canismarko) [inline apstools/devices/area_detector_factory.py:16]: *plugins* needs to be a keyword argument for this to work.

File: apstools/devices/area_detector_factory.py [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+
+View ADSimDetector image with CA and PVA::
+
+    from ophyd.areadetector import SimDetectorCam
+    from apstools.devices import ad_creator
+
+    det = ad_creator(
+        "ad:", name="det", class_name="MySimDetector",
+        [
```
Comment (canismarko) [inline apstools/devices/area_detector_factory.py:25]: Make a kw argument

File: apstools/devices/area_detector_factory.py [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+
+Record HDF5 images with Eiger detector. Here, both the Eiger detector IOC and
+the Bluesky databroker use the same filesystem mount ``/``::
+
+    from ophyd.areadetector import EigerDetectorCam
+    from apstools.devices import ad_creator
+
+    det = ad_creator(
+        "ad:", name="det", class_name"MyEiger",
+        [
```
Comment (canismarko) [inline apstools/devices/area_detector_factory.py:40]: Make a kw argument.

File: apstools/devices/area_detector_factory.py [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+
+    from ophyd.areadetector import EigerDetectorCam
+    from apstools.devices import ad_creator, PLUGIN_DEFAULTS
+
+    plugin_defaults = PLUGIN_DEFAULTS.copy()
+    plugin_defaults["hdf1"].pop("read_path_template", None)
+    plugin_defaults["hdf1"].pop("write_path_template", None)
+
+    det = ad_creator(
+        "ad:", name="det", class_name"MyEiger",
```
Comment (canismarko) [inline apstools/devices/area_detector_factory.py:59]: Make into kw arguments.

**Review by canismarko — CHANGES_REQUESTED** [review]
Updating my reivew based on Pete's point that these actually need to be keyword arguments in the examples.

File: apstools/devices/tests/test_issue984_ad_factory.py
Comment (canismarko) [reply apstools/devices/tests/test_issue984_ad_factory.py:1]: Ah yes, I think you would get a "positional argument follows keyword argument" error as-is. I updated my review.

File: apstools/devices/area_detector_factory.py (hunk 5 above)
Comment (prjemian) [reply apstools/devices/area_detector_factory.py:59]: This is the most problematic of the `plugin_defaults` list structure.  It won't work as written since `plugin_defaults` is a Python `list`.  Which means a unit test should be provided.

Comment (canismarko) [reply apstools/devices/area_detector_factory.py:59]: ``PLUGIN_DEFAULTS`` is a dict (https://github.com/BCDA-APS/apstools/blob/61c93bbc81a67b46d61855a57433a6756c20831e/apstools/devices/area_detector_factory.py#L82), and *plugin_defaults* is type-hinted and used as a dict.

Comment (prjemian) [reply apstools/devices/area_detector_factory.py:59]: Right.  Exactly _that_ kind of confusion.

Comment (prjemian) [reply apstools/devices/area_detector_factory.py:59]: Example is correct as written, once the `plugins=` is added.

File: apstools/devices/area_detector_factory.py [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+        *object*:
+        Plugin configuration dictionary.
+        (default: ``None``, PLUGIN_DEFAULTS will be used.)
+    """
+    bases = bases or (
+        SingleTrigger_V34,
+        ophyd.areadetector.DetectorBase,
+    )
+    plugins = plugins or ["cam"]
+    plugin_defaults = plugin_defaults or PLUGIN_DEFAULTS
```
Comment (canismarko) [inline apstools/devices/area_detector_factory.py:244]: This makes it impossible to pass an empty dict (``{}``) to this function as *plugin_defaults*, since ``bool({})`` evaluates to ``False`` and then ``PLUGIN_DEFAULTS`` would get used.

Maybe not a big deal since I can't imagine many people would want this. But I think it could be avoided by using ``PLUGIN_DEFAULTS`` as the default value for the *plugin_defaults*, which makes the intention more clear when looking at the call signature.

Comment (prjemian) [reply apstools/devices/area_detector_factory.py:244]: Can do.

Comment (prjemian) [reply apstools/devices/area_detector_factory.py:244]: Same for `plugins=None`.  Changing it to `plugins=["cam"]`.

Comment (prjemian) [reply apstools/devices/area_detector_factory.py:244]: Ah... there's the reason to keep the default kwarg value of `None` -- in the unit tests, `None` is used.

```py
ERROR apstools/devices/tests/test_area_detector_support.py::test_AD_EpicsFileNameMixin[hdf1-AD_HDF5] - TypeError: Must be a dict.  Received plugin_defaults=None
ERROR apstools/devices/tests/test_area_detector_support.py::test_AD_EpicsFileNameMixin[jpeg1-AD_JPEG] - TypeError: Must be a dict.  Received plugin_defaults=None
ERROR apstools/devices/tests/test_area_detector_support.py::test_AD_EpicsFileNameMixin[tiff1-AD_TIFF] - TypeError: Must be a dict.  Received plugin_defaults=None
ERROR apstools/devices/tests/test_area_detector_support.py::test_stage[hdf1-AD_EpicsFileNameHDF5Plugin] - TypeError: Must be a dict.  Received plugin_defaults=None
ERROR apstools/devices/tests/test_area_detector_support.py::test_stage[jpeg1-AD_EpicsFileNameJPEGPlugin] - TypeError: Must be a dict.  Received plugin_defaults=None
ERROR apstools/devices/tests/test_area_detector_support.py::test_stage[tiff1-AD_EpicsFileNameTIFFPlugin] - TypeError: Must be a dict.  Received plugin_defaults=None
ERROR apstools/devices/tests/test_area_detector_support.py::test_bp_count_custom_name[hdf1-AD_EpicsFileNameHDF5Plugin] - TypeError: Must be a dict.  Received plugin_defaults=None
ERROR apstools/devices/tests/test_area_detector_support.py::test_bp_count_custom_name[jpeg1-AD_EpicsFileNameJPEGPlugin] - TypeError: Must be a dict.  Received plugin_defaults=None
ERROR apstools/devices/tests/test_area_detector_support.py::test_bp_count_custom_name[tiff1-AD_EpicsFileNameTIFFPlugin] - TypeError: Must be a dict.  Received plugin_defaults=None
ERROR apstools/devices/tests/test_area_detector_support.py::test_full_file_name_local - TypeError: Must be a dict.  Received plugin_defaults=None
ERROR apstools/devices/tests/test_area_detector_support.py::test_no_file_path - TypeError: Must be a dict.  Received plugin_defaults=None
ERROR apstools/devices/tests/test_area_detector_support.py::test_file_numbering - TypeError: Must be a dict.  Received plugin_defaults=None
ERROR apstools/devices/tests/test_area_detector_support.py::test_capture_error_with_single_mode - TypeError: Must be a dict.  Received plugin_defaults=None
ERROR apstools/devices/tests/test_area_detector_support.py::test_single_mode - TypeError: type.__new__() argument 2 must be tuple, not None
```

Code like this is needed:

```py
    if plugin_defaults is None:
        plugin_defaults = PLUGIN_DEFAULTS
```

Comment (canismarko) [reply apstools/devices/area_detector_factory.py:244]: Okay, sounds good.

Comment (prjemian) [reply apstools/devices/area_detector_factory.py:244]: > I can't imagine many people would want `plugin_defaults={}`

Actually, this is a realistic case to force the caller to define all keywords of each plugin.  Probably only useful for unit testing, not real world?

Comment (prjemian) [reply apstools/devices/area_detector_factory.py]: @canismarko Will these changes work for you?

**Review by canismarko — APPROVED** [review]

File: apstools/devices/area_detector_factory.py (hunk 6 above)
Comment (canismarko) [reply apstools/devices/area_detector_factory.py]: Yes, changes look good to me. Thanks!

Comment (prjemian) [reply apstools/devices/area_detector_factory.py]: I also added some documentation (look for `EXAMPLE 1:` and others).
